### PR TITLE
sending lab request to ALIS

### PR DESCRIPTION
**Pull request by Mwanje**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/ugandaemrsync/server/SyncConstant.java [hunk 1]
```diff
@@ -71,6 +71,7 @@
     public static final String VIRAL_LOAD_SYNC_TYPE_UUID = "3551ca84-06c0-432b-9064-fcfeefd6f4ec";
     public static final String FHIRSERVER_SYNC_TASK_TYPE_UUID = "3c1ce940-8ade-11ea-bc55-0242ac130003";
     public static final String VIRAL_LOAD_RESULT_PULL_TYPE_UUID = "3396dcf0-2106-4e73-9b90-c63978c3a8b4";
+	public static final String OTHER_TESTS_SYNC_TYPE_UUID = "64078546-cd3a-4669-96a4-a8d0722b7be7";
```
[the remark this answers is not in the document]
Comment (Mwanje) [reply api/src/main/java/org/openmrs/module/ugandaemrsync/server/SyncConstant.java:74]: Done

File: api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/SendLabRequestToALIS.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        } catch (ParseException e) {
+            log.error("Failed to pass orders to list", e);
+        }
+
+        SyncTaskType syncTaskType = ugandaEMRSyncService.getSyncTaskTypeByUUID(OTHER_TESTS_SYNC_TYPE_UUID);
+
+        for (Order order : orderList) {
+            SyncTask syncTask = ugandaEMRSyncService.getSyncTaskBySyncTaskId(order.getAccessionNumber());
+            if (syncTask == null) {
+                Map<String, String> dataOutput = generateLabFHIROrderTestRequestBody((TestOrder) order, VL_SEND_SAMPLE_FHIR_JSON_STRING);
```
[the remark this answers is not in the document]
Comment (Mwanje) [reply api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/SendLabRequestToALIS.java:53]: Done

File: api/src/main/java/org/openmrs/module/ugandaemrsync/server/SyncConstant.java [hunk 3]
```diff
@@ -452,6 +453,8 @@
 
     public static final String VIRAL_LOAD_ORDERS_QUERY = "select orders.order_id from orders  inner join test_order on (test_order.order_id=orders.order_id) inner JOIN concept ON (orders.concept_id = concept.concept_id) inner join concept_reference_map on (concept.concept_id = concept_reference_map.concept_id) inner join concept_reference_term on (concept_reference_map.concept_reference_term_id = concept_reference_term.concept_reference_term_id) where accession_number!=\"\" AND specimen_source!=\"\" AND orders.instructions=\"REFER TO cphl\" AND code=\"" + VIRAL_LOAD_SYNC_TASK_TYPE_IDENTIFIER + "\"";
 
+	public static final String LAB_ORDERS_QUERY = "select orders.order_id from orders  inner join test_order on (test_order.order_id=orders.order_id) inner JOIN concept ON (orders.concept_id = concept.concept_id) inner join concept_reference_map on (concept.concept_id = concept_reference_map.concept_id) inner join concept_reference_term on (concept_reference_map.concept_reference_term_id = concept_reference_term.concept_reference_term_id) where accession_number!=\"\" AND specimen_source!=\"\"";
```
[the remark this answers is not in the document]
Comment (Mwanje) [reply api/src/main/java/org/openmrs/module/ugandaemrsync/server/SyncConstant.java:456]: Done

File: api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/SendLabRequestToALIS.java [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    protected Log log = LogFactory.getLog(SendLabRequestToALIS.class);
+
+    @Override
+    public void execute() {
+        UgandaEMRHttpURLConnection ugandaEMRHttpURLConnection = new UgandaEMRHttpURLConnection();
+        UgandaEMRSyncService ugandaEMRSyncService = Context.getService(UgandaEMRSyncService.class);
+        List<Order> orderList = new ArrayList<>();
+
+        if (!ugandaEMRHttpURLConnection.isConnectionAvailable()) {
+            return;
```
Comment (dbaluku) [inline api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/SendLabRequestToALIS.java:37]: @Mwanje we also need to first check whether the server is reachable before we process the orders

File: api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/SendLabRequestToALIS.java [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+        } catch (IOException e) {
+            log.error("Failed to get orders", e);
+        } catch (ParseException e) {
+            log.error("Failed to pass orders to list", e);
+        }
+
+        SyncTaskType syncTaskType = ugandaEMRSyncService.getSyncTaskTypeByUUID(ALIS_TESTS_SYNC_TYPE_UUID);
+
+        for (Order order : orderList) {
```
Comment (dbaluku) [inline api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/SendLabRequestToALIS.java:52]: @Mwanje  i think we need a check to ensure the orderlist is not null and empty

File: api/src/main/java/org/openmrs/module/ugandaemrsync/api/UgandaEMRSyncService.java [hunk 6]
```diff
@@ -9,22 +9,17 @@
  */
 package org.openmrs.module.ugandaemrsync.api;
 
-import org.openmrs.Encounter;
-import org.openmrs.EncounterType;
-import org.openmrs.Patient;
-import org.openmrs.Order;
-import org.openmrs.annotation.Authorized;
+import org.openmrs.*;
```
Comment (dbaluku) [inline api/src/main/java/org/openmrs/module/ugandaemrsync/api/UgandaEMRSyncService.java:12]: Removed wild card

File: api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/SendLabRequestToALIS.java [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+            SyncTask syncTask = ugandaEMRSyncService.getSyncTaskBySyncTaskId(order.getAccessionNumber());
+            if (syncTask == null) {
+                Map<String, String> dataOutput = generateLabFHIROrderTestRequestBody((TestOrder) order, ALIS_SEND_SAMPLE_FHIR_JSON_STRING);
+                String json = dataOutput.get("json");
+                try {
+                    Map map = ugandaEMRHttpURLConnection.sendPostBy(syncTaskType.getUrl(), syncTaskType.getUrlUserName(), syncTaskType.getUrlPassword(), "", json, false);
```
Comment (dbaluku) [inline api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/SendLabRequestToALIS.java:58]: This line throws some null point exception, i guess one the parameters has possible value as null

File: api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/SendLabRequestToALIS.java (hunk 5 above)
Comment (Mwanje) [reply api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/SendLabRequestToALIS.java:52]: Done

File: api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/SendLabRequestToALIS.java (hunk 4 above)
Comment (Mwanje) [reply api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/SendLabRequestToALIS.java:37]: Done

**Pull request closed without merging**